<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Info-about-my-dataset" data-toc-modified-id="Info-about-my-dataset-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Info about my dataset</a></span><ul class="toc-item"><li><span><a href="#Cleaning-the-race-column" data-toc-modified-id="Cleaning-the-race-column-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Cleaning the race column</a></span></li></ul></li><li><span><a href="#Understand-the-features" data-toc-modified-id="Understand-the-features-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Understand the features</a></span></li><li><span><a href="#MVP:-Logistic-Regression" data-toc-modified-id="MVP:-Logistic-Regression-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>MVP: Logistic Regression</a></span></li></ul></div>

# Info about my dataset

In [133]:
import pandas as pd
data = pd.read_csv("/Users/fatimaloumaini/Downloads/dataset_diabetes/diabetic_data.csv")

In [134]:
# Data dictionary 
data_dict = pd.read_csv("/Users/fatimaloumaini/Downloads/dataset_diabetes/IDs_mapping.csv")

In [135]:
data.head()

encounter_id  patient_nbr             race  gender      age weight  \
0       2278392      8222157        Caucasian  Female   [0-10)      ?   
1        149190     55629189        Caucasian  Female  [10-20)      ?   
2         64410     86047875  AfricanAmerican  Female  [20-30)      ?   
3        500364     82442376        Caucasian    Male  [30-40)      ?   
4         16680     42519267        Caucasian    Male  [40-50)      ?   

   admission_type_id  discharge_disposition_id  admission_source_id  \
0                  6                        25                    1   
1                  1                         1                    7   
2                  1                         1                    7   
3                  1                         1                    7   
4                  1                         1                    7   

   time_in_hospital  ... citoglipton insulin  glyburide-metformin  \
0                 1  ...          No      No                   No   
1                 3  ...          No      Up                   No   
2                 2  ...          No      No                   No   
3                 2  ...          No      Up                   No   
4                 1  ...          No  Steady                   No   

   glipizide-metformin  glimepiride-pioglitazone  metformin-rosiglitazone  \
0                   No                        No                       No   
1                   No                        No                       No   
2                   No                        No                       No   
3                   No                        No                       No   
4                   No                        No                       No   

   metformin-pioglitazone  change diabetesMed readmitted  
0                      No      No          No         NO  
1                      No      Ch         Yes        >30  
2                      No      No         Yes         NO  
3                      No      Ch         Yes         NO  
4                      No      Ch         Yes         NO  

[5 rows x 50 columns]

In [136]:
data.describe()

encounter_id   patient_nbr  admission_type_id  \
count  1.017660e+05  1.017660e+05      101766.000000   
mean   1.652016e+08  5.433040e+07           2.024006   
std    1.026403e+08  3.869636e+07           1.445403   
min    1.252200e+04  1.350000e+02           1.000000   
25%    8.496119e+07  2.341322e+07           1.000000   
50%    1.523890e+08  4.550514e+07           1.000000   
75%    2.302709e+08  8.754595e+07           3.000000   
max    4.438672e+08  1.895026e+08           8.000000   

       discharge_disposition_id  admission_source_id  time_in_hospital  \
count             101766.000000        101766.000000     101766.000000   
mean                   3.715642             5.754437          4.395987   
std                    5.280166             4.064081          2.985108   
min                    1.000000             1.000000          1.000000   
25%                    1.000000             1.000000          2.000000   
50%                    1.000000             7.000000          4.000000   
75%                    4.000000             7.000000          6.000000   
max                   28.000000            25.000000         14.000000   

       num_lab_procedures  num_procedures  num_medications  number_outpatient  \
count       101766.000000   101766.000000    101766.000000      101766.000000   
mean            43.095641        1.339730        16.021844           0.369357   
std             19.674362        1.705807         8.127566           1.267265   
min              1.000000        0.000000         1.000000           0.000000   
25%             31.000000        0.000000        10.000000           0.000000   
50%             44.000000        1.000000        15.000000           0.000000   
75%             57.000000        2.000000        20.000000           0.000000   
max            132.000000        6.000000        81.000000          42.000000   

       number_emergency  number_inpatient  number_diagnoses  
count     101766.000000     101766.000000     101766.000000  
mean           0.197836          0.635566          7.422607  
std            0.930472          1.262863          1.933600  
min            0.000000          0.000000          1.000000  
25%            0.000000          0.000000          6.000000  
50%            0.000000          0.000000          8.000000  
75%            0.000000          1.000000          9.000000  
max           76.000000         21.000000         16.000000

In [137]:
data.columns

Index(['encounter_id', 'patient_nbr', 'race', 'gender', 'age', 'weight',
       'admission_type_id', 'discharge_disposition_id', 'admission_source_id',
       'time_in_hospital', 'payer_code', 'medical_specialty',
       'num_lab_procedures', 'num_procedures', 'num_medications',
       'number_outpatient', 'number_emergency', 'number_inpatient', 'diag_1',
       'diag_2', 'diag_3', 'number_diagnoses', 'max_glu_serum', 'A1Cresult',
       'metformin', 'repaglinide', 'nateglinide', 'chlorpropamide',
       'glimepiride', 'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide',
       'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone',
       'tolazamide', 'examide', 'citoglipton', 'insulin',
       'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted'],
      dtype='object')

In [139]:
for col in data.columns:
    if data[col].dtype == object:
         print(col,data[col][data[col] == '?'].count())
# gender was coded differently so we use a custom count for this one
#print('gender', test['gender'][test['gender'] == 'Unknown/Invalid'].count())

race 2273
gender 0
age 0
weight 98569
payer_code 40256
medical_specialty 49949
diag_1 21
diag_2 358
diag_3 1423
max_glu_serum 0
A1Cresult 0
metformin 0
repaglinide 0
nateglinide 0
chlorpropamide 0
glimepiride 0
acetohexamide 0
glipizide 0
glyburide 0
tolbutamide 0
pioglitazone 0
rosiglitazone 0
acarbose 0
miglitol 0
troglitazone 0
tolazamide 0
examide 0
citoglipton 0
insulin 0
glyburide-metformin 0
glipizide-metformin 0
glimepiride-pioglitazone 0
metformin-rosiglitazone 0
metformin-pioglitazone 0
change 0
diabetesMed 0
readmitted 0


In [140]:
data.shape

(101766, 50)

- race 2273: check what I will do with it
- weight 98569: drop this column 96% missing values
- payer_code 40256: drop
- - medical_specialty 49949: drop
- diag_1 21: : check what I will do with it
- diag_2 358: : check what I will do with it
- diag_3 1423: check what I will do with it

In [141]:
data = data.drop(["weight","payer_code", "medical_specialty"], axis = 1)

In [142]:
data.shape

(101766, 47)

In [143]:
data= data[(data['diag_1'] != '?') & (data['diag_2'] != '?') & (data['diag_3'] != '?')]


In [144]:
# It means the data is good to stay
data.shape

(100244, 47)

In [145]:
data.head()

encounter_id  patient_nbr             race  gender      age  \
1        149190     55629189        Caucasian  Female  [10-20)   
2         64410     86047875  AfricanAmerican  Female  [20-30)   
3        500364     82442376        Caucasian    Male  [30-40)   
4         16680     42519267        Caucasian    Male  [40-50)   
5         35754     82637451        Caucasian    Male  [50-60)   

   admission_type_id  discharge_disposition_id  admission_source_id  \
1                  1                         1                    7   
2                  1                         1                    7   
3                  1                         1                    7   
4                  1                         1                    7   
5                  2                         1                    2   

   time_in_hospital  num_lab_procedures  ...  citoglipton  insulin  \
1                 3                  59  ...           No       Up   
2                 2                  11  ...           No       No   
3                 2                  44  ...           No       Up   
4                 1                  51  ...           No   Steady   
5                 3                  31  ...           No   Steady   

   glyburide-metformin  glipizide-metformin  glimepiride-pioglitazone  \
1                   No                   No                        No   
2                   No                   No                        No   
3                   No                   No                        No   
4                   No                   No                        No   
5                   No                   No                        No   

  metformin-rosiglitazone metformin-pioglitazone change  diabetesMed  \
1                      No                     No     Ch          Yes   
2                      No                     No     No          Yes   
3                      No                     No     Ch          Yes   
4                      No                     No     Ch          Yes   
5                      No                     No     No          Yes   

  readmitted  
1        >30  
2         NO  
3         NO  
4         NO  
5        >30  

[5 rows x 47 columns]

In [146]:
for col in data.columns:
    if data[col].dtype == object:
         print(col,data[col][data[col] == '?'].count())

race 2191
gender 0
age 0
diag_1 0
diag_2 0
diag_3 0
max_glu_serum 0
A1Cresult 0
metformin 0
repaglinide 0
nateglinide 0
chlorpropamide 0
glimepiride 0
acetohexamide 0
glipizide 0
glyburide 0
tolbutamide 0
pioglitazone 0
rosiglitazone 0
acarbose 0
miglitol 0
troglitazone 0
tolazamide 0
examide 0
citoglipton 0
insulin 0
glyburide-metformin 0
glipizide-metformin 0
glimepiride-pioglitazone 0
metformin-rosiglitazone 0
metformin-pioglitazone 0
change 0
diabetesMed 0
readmitted 0


## Cleaning the race column

In [147]:
data["race"].value_counts()

Caucasian          75079
AfricanAmerican    18881
?                   2191
Hispanic            1984
Other               1484
Asian                625
Name: race, dtype: int64

In [148]:

data[data["race"] == "?"]

encounter_id  patient_nbr race  gender       age  admission_type_id  \
19            148530     69422211    ?    Male   [70-80)                  3   
20            150006     22864131    ?  Female   [50-60)                  2   
21            150048     21239181    ?    Male   [60-70)                  2   
54            604188    113848434    ?  Female   [60-70)                  3   
65            713268     58231314    ?  Female  [90-100)                  2   
87           1139226     73921680    ?    Male   [60-70)                  2   
99           1260216     92117574    ?    Male   [30-40)                  2   
111          1686690     10777860    ?    Male   [60-70)                  3   
127          2139660     18030078    ?    Male   [30-40)                  2   
177          2562210      8005509    ?    Male   [40-50)                  6   
183          2570172       357300    ?    Male   [50-60)                  6   
207          2677470      4943709    ?  Female   [80-90)                  6   
238          2852430      2341521    ?  Female   [70-80)                  6   
407          3775698     29173437    ?  Female   [80-90)                  3   
442          3973368     55890810    ?    Male   [60-70)                  3   
480          4174734       707769    ?  Female   [60-70)                  6   
522          4373364       886266    ?    Male   [60-70)                  6   
546          4493220     92948409    ?  Female   [60-70)                  2   
580          4682388       707769    ?  Female   [60-70)                  6   
643          5053842      1041291    ?    Male   [50-60)                  6   
645          5074566     22829274    ?    Male   [50-60)                  1   
652          5123124      2535174    ?    Male   [50-60)                  6   
664          5231520     64511352    ?    Male   [60-70)                  1   
721          5516124     22783734    ?    Male   [50-60)                  2   
732          5652390     27926721    ?  Female   [50-60)                  2   
742          5683854     56222487    ?    Male   [60-70)                  1   
813          6392556     64603215    ?    Male   [40-50)                  2   
836          6572046     41275089    ?    Male   [50-60)                  3   
912          6993210     52728858    ?    Male   [60-70)                  1   
927          7125354      9402138    ?    Male   [30-40)                  6   
...              ...          ...  ...     ...       ...                ...   
100951     433104848    110697390    ?    Male   [70-80)                  1   
100975     433212602     38464254    ?    Male   [60-70)                  1   
100998     433713650    129794522    ?    Male   [60-70)                  1   
101084     434891720     38254788    ?  Female   [50-60)                  1   
101089     434901176     47013921    ?    Male   [80-90)                  1   
101091     434901332     76265874    ?  Female   [70-80)                  2   
101218     436656206     38886651    ?    Male   [30-40)                  1   
101227     436694090     98835705    ?  Female  [90-100)                  3   
101230     436694174     49166001    ?  Female   [40-50)                  1   
101257     437219804     67711356    ?    Male   [30-40)                  1   
101263     437243192     38090619    ?  Female  [90-100)                  3   
101280     437313266    120476120    ?  Female   [80-90)                  1   
101289     437338094    132239066    ?    Male   [60-70)                  2   
101294     437362598     59545791    ?    Male   [80-90)                  1   
101352     438424970    131369585    ?    Male   [70-80)                  1   
101362     438453374    126546161    ?    Male   [50-60)                  1   
101400     438573830    153620564    ?  Female   [70-80)                  2   
101424     439040900     39973257    ?    Male   [70-80)                  1   
101445     439160162     38321100    ?  Female   [70-80)  

Since race is an important feature I will delete the row with ?

In [149]:
new_df = data[data["race"] != "?"]

In [150]:
new_df.shape

(100244, 47)

In [151]:
new_df["readmitted"].value_counts()

NO     52338
>30    34649
<30    11066
Name: readmitted, dtype: int64

Since I am looking to predict the readmission rate to a patient, it makes sense to remove the deceased patients and that correspond to 

In [152]:
new_df= new_df[new_df["discharge_disposition_id"] != 11]

In [153]:
new_df.shape

(96447, 47)

In [154]:
new_df.head()

encounter_id  patient_nbr             race  gender      age  \
1        149190     55629189        Caucasian  Female  [10-20)   
2         64410     86047875  AfricanAmerican  Female  [20-30)   
3        500364     82442376        Caucasian    Male  [30-40)   
4         16680     42519267        Caucasian    Male  [40-50)   
5         35754     82637451        Caucasian    Male  [50-60)   

   admission_type_id  discharge_disposition_id  admission_source_id  \
1                  1                         1                    7   
2                  1                         1                    7   
3                  1                         1                    7   
4                  1                         1                    7   
5                  2                         1                    2   

   time_in_hospital  num_lab_procedures  ...  citoglipton  insulin  \
1                 3                  59  ...           No       Up   
2                 2                  11  ...           No       No   
3                 2                  44  ...           No       Up   
4                 1                  51  ...           No   Steady   
5                 3                  31  ...           No   Steady   

   glyburide-metformin  glipizide-metformin  glimepiride-pioglitazone  \
1                   No                   No                        No   
2                   No                   No                        No   
3                   No                   No                        No   
4                   No                   No                        No   
5                   No                   No                        No   

  metformin-rosiglitazone metformin-pioglitazone change  diabetesMed  \
1                      No                     No     Ch          Yes   
2                      No                     No     No          Yes   
3                      No                     No     Ch          Yes   
4                      No                     No     Ch          Yes   
5                      No                     No     No          Yes   

  readmitted  
1        >30  
2         NO  
3         NO  
4         NO  
5        >30  

[5 rows x 47 columns]

In [155]:
new_df.columns

Index(['encounter_id', 'patient_nbr', 'race', 'gender', 'age',
       'admission_type_id', 'discharge_disposition_id', 'admission_source_id',
       'time_in_hospital', 'num_lab_procedures', 'num_procedures',
       'num_medications', 'number_outpatient', 'number_emergency',
       'number_inpatient', 'diag_1', 'diag_2', 'diag_3', 'number_diagnoses',
       'max_glu_serum', 'A1Cresult', 'metformin', 'repaglinide', 'nateglinide',
       'chlorpropamide', 'glimepiride', 'acetohexamide', 'glipizide',
       'glyburide', 'tolbutamide', 'pioglitazone', 'rosiglitazone', 'acarbose',
       'miglitol', 'troglitazone', 'tolazamide', 'examide', 'citoglipton',
       'insulin', 'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted'],
      dtype='object')

# Understand the features

- encouter_id: Unique identifier of an encounter
- patient_nbr: Unique identifier of a patient
- race: Values: Caucasian, Asian, African American, Hispanic, and other
- gender: Values: male, female, and unknown/invalid
- age: Grouped in 10-year intervals:[0, 10), [10, 20), . . ., [90, 10]
- admission_type_id: Integer identifier corresponding to 8 distinct values
- discharge_disposition_id: Integer identifier corresponding to 29 distinct values
- admission_source_id: Integer identifier corresponding to 21 distinct values
- time in hospital
- num_lab_procedures: Number of lab tests performed during the encounter
- num_procedures: Number of procedures (other than lab tests) performed during the encounter
- num_medications: Number of distinct generic names administered during the encounter
- number_outpatient(No admission to the hospital): Number of outpatient visits of the patient in the year preceding the encounter
- number_emergency: Number of emergency visits of the patient in the year preceding the encounter
- number_inpatient(admitted to the hospital): Number of inpatient visits of the patient in the year preceding the encounter
- diag_1: The primary diagnosis (coded as first three digits of ICD9); 848 distinct values
- diag_2: Secondary diagnosis (coded as first three digits of ICD9); 923 distinct values
- diag_3: Additional secondary diagnosis (coded as first three digits of ICD9); 954 distinct values
- number_diagnoses: Number of diagnoses entered to the system
- max_glu_serum: Glucose serum test : Indicates the range of the result or if the test was not taken. Values: “>200,” “>300,” “normal,” and “none” if not measured
- A1Cresult( alc test result): was greater than 8%, “>7” if the result was greater than 7% but less than 8%, “normal” if the result was less than 7%, and “none” if not measured.
- change: Indicates if there was a change in diabetic medications (either dosage or generic name). Values: “change” and “no change”
- features for medication: the feature indicates whether the drug was prescribed or there was a change in the dosage. Values: “up” if the dosage was increased during the encounter, “down” if the dosage was decreased, “steady” if the dosage did not change, and “no” if the drug was not prescribed.
- diabetes medications: Indicates if there was any diabetic medication prescribed. Values: “yes” and “no”
For the generic names: metformin, repaglinide, nateglinide, chlorpropamide, glimepiride, acetohexamide, glipizide, glyburide, tolbutamide, pioglitazone, rosiglitazone, acarbose, miglitol, troglitazone, tolazamide, examide, sitagliptin, insulin, glyburide-metformin, glipizide-metformin, glimepiride-pioglitazone, metformin-rosiglitazone, and metformin-pioglitazone, the feature indicates whether the drug was prescribed or there was a change in the dosage. 


In [156]:
new_df["glimepiride"].value_counts()

No        91501
Steady     4451
Up          312
Down        183
Name: glimepiride, dtype: int64

Medications is either No, steady, Up, down which means that they represent the change in the dosage. I want to know how many changes are made for each patient

In [157]:
keys = ['metformin', 'repaglinide', 'nateglinide', 'chlorpropamide', 'glimepiride', 'glipizide', 'glyburide', 'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'insulin', 'glyburide-metformin', 'tolazamide', 'metformin-pioglitazone','metformin-rosiglitazone', 'glimepiride-pioglitazone', 'glipizide-metformin', 'troglitazone', 'tolbutamide', 'acetohexamide']
for col in keys:
    colname = str(col) + 'temp'
    new_df[colname] = new_df[col].apply(lambda x: 0 if (x == 'No' or x == 'Steady') else 1)
new_df['dosage_change'] = 0
for col in keys:
    colname = str(col) + 'temp'
    new_df['dosage_change'] = new_df['dosage_change'] + new_df[colname]
    del new_df[colname]

In [158]:
new_df.head()


encounter_id  patient_nbr             race  gender      age  \
1        149190     55629189        Caucasian  Female  [10-20)   
2         64410     86047875  AfricanAmerican  Female  [20-30)   
3        500364     82442376        Caucasian    Male  [30-40)   
4         16680     42519267        Caucasian    Male  [40-50)   
5         35754     82637451        Caucasian    Male  [50-60)   

   admission_type_id  discharge_disposition_id  admission_source_id  \
1                  1                         1                    7   
2                  1                         1                    7   
3                  1                         1                    7   
4                  1                         1                    7   
5                  2                         1                    2   

   time_in_hospital  num_lab_procedures  ...  insulin  glyburide-metformin  \
1                 3                  59  ...       Up                   No   
2                 2                  11  ...       No                   No   
3                 2                  44  ...       Up                   No   
4                 1                  51  ...   Steady                   No   
5                 3                  31  ...   Steady                   No   

   glipizide-metformin  glimepiride-pioglitazone  metformin-rosiglitazone  \
1                   No                        No                       No   
2                   No                        No                       No   
3                   No                        No                       No   
4                   No                        No                       No   
5                   No                        No                       No   

  metformin-pioglitazone change diabetesMed  readmitted dosage_change  
1                     No     Ch         Yes         >30             1  
2                     No     No         Yes          NO             0  
3                     No     Ch         Yes          NO             1  
4                     No     Ch         Yes          NO             0  
5                     No     No         Yes         >30             0  

[5 rows x 48 columns]

In [159]:
new_df["dosage_change"].value_counts()

0    70143
1    24922
2     1271
3      106
4        5
Name: dosage_change, dtype: int64

In [168]:
#Recoding the target variable
new_df['readmitted'] = new_df['readmitted'].replace('>30', 0)
new_df['readmitted'] = new_df['readmitted'].replace('<30', 1)
new_df['readmitted'] = new_df['readmitted'].replace('NO', 0)

In [169]:
new_df["readmitted"].value_counts()


0    85381
1    11066
Name: readmitted, dtype: int64

In [176]:
new_df.head()

encounter_id  patient_nbr             race  gender      age  \
1        149190     55629189        Caucasian  Female  [10-20)   
2         64410     86047875  AfricanAmerican  Female  [20-30)   
3        500364     82442376        Caucasian    Male  [30-40)   
4         16680     42519267        Caucasian    Male  [40-50)   
5         35754     82637451        Caucasian    Male  [50-60)   

   admission_type_id  discharge_disposition_id  admission_source_id  \
1                  1                         1                    7   
2                  1                         1                    7   
3                  1                         1                    7   
4                  1                         1                    7   
5                  2                         1                    2   

   time_in_hospital  num_lab_procedures  ...  insulin  glyburide-metformin  \
1                 3                  59  ...       Up                   No   
2                 2                  11  ...       No                   No   
3                 2                  44  ...       Up                   No   
4                 1                  51  ...   Steady                   No   
5                 3                  31  ...   Steady                   No   

   glipizide-metformin  glimepiride-pioglitazone  metformin-rosiglitazone  \
1                   No                        No                       No   
2                   No                        No                       No   
3                   No                        No                       No   
4                   No                        No                       No   
5                   No                        No                       No   

  metformin-pioglitazone change diabetesMed  readmitted dosage_change  
1                     No     Ch         Yes           0             1  
2                     No     No         Yes           0             0  
3                     No     Ch         Yes           0             1  
4                     No     Ch         Yes           0             0  
5                     No     No         Yes           0             0  

[5 rows x 48 columns]

In [178]:
new_df.columns

Index(['encounter_id', 'patient_nbr', 'race', 'gender', 'age',
       'admission_type_id', 'discharge_disposition_id', 'admission_source_id',
       'time_in_hospital', 'num_lab_procedures', 'num_procedures',
       'num_medications', 'number_outpatient', 'number_emergency',
       'number_inpatient', 'diag_1', 'diag_2', 'diag_3', 'number_diagnoses',
       'max_glu_serum', 'A1Cresult', 'metformin', 'repaglinide', 'nateglinide',
       'chlorpropamide', 'glimepiride', 'acetohexamide', 'glipizide',
       'glyburide', 'tolbutamide', 'pioglitazone', 'rosiglitazone', 'acarbose',
       'miglitol', 'troglitazone', 'tolazamide', 'examide', 'citoglipton',
       'insulin', 'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted',
       'dosage_change'],
      dtype='object')

In [179]:
new_df["age"].value_counts()

[70-80)     24816
[60-70)     21521
[50-60)     16546
[80-90)     16223
[40-50)      9208
[30-40)      3538
[90-100)     2594
[20-30)      1471
[10-20)       466
[0-10)         64
Name: age, dtype: int64

In [182]:
# Change the age to numerical values

age_dict = {"[0-10)":5, "[10-20)":15, "[20-30)":25, "[30-40)":35, "[40-50)":45, "[50-60)":55, "[60-70)":65, "[70-80)":75, "[80-90)":85, "[90-100)":95}
new_df['age'] = new_df.age.map(age_dict)
new_df['age'] = new_df['age'].astype('int64')

encounter_id   patient_nbr           age  admission_type_id  \
count  9.644700e+04  9.644700e+04  96447.000000       96447.000000   
mean   1.659535e+08  5.483692e+07     66.127044           2.030514   
std    1.025184e+08  3.868106e+07     15.594280           1.450935   
min    1.252200e+04  1.350000e+02      5.000000           1.000000   
25%    8.522830e+07  2.348840e+07     55.000000           1.000000   
50%    1.534117e+08  4.690138e+07     65.000000           1.000000   
75%    2.308025e+08  8.804246e+07     75.000000           3.000000   
max    4.438672e+08  1.895026e+08     95.000000           8.000000   

       discharge_disposition_id  admission_source_id  time_in_hospital  \
count              96447.000000         96447.000000      96447.000000   
mean                   3.632700             5.760563          4.415420   
std                    5.269729             4.074101          2.982335   
min                    1.000000             1.000000          1.000000   
25%                    1.000000             1.000000          2.000000   
50%                    1.000000             7.000000          4.000000   
75%                    3.000000             7.000000          6.000000   
max                   28.000000            25.000000         14.000000   

       num_lab_procedures  num_procedures  num_medications  number_outpatient  \
count        96447.000000    96447.000000     96447.000000       96447.000000   
mean            42.992763        1.342074        16.080282           0.376497   
std             19.657060        1.703215         8.072497           1.280055   
min              1.000000        0.000000         1.000000           0.000000   
25%             31.000000        0.000000        11.000000           0.000000   
50%             44.000000        1.000000        15.000000           0.000000   
75%             57.000000        2.000000        20.000000           0.000000   
max            132.000000        6.000000        81.000000          42.000000   

       number_emergency  number_inpatient  number_diagnoses    readmitted  \
count      96447.000000      96447.000000      96447.000000  96447.000000   
mean           0.202992          0.644126          7.499995      0.114737   
std            0.948084          1.269970          1.836684      0.318705   
min            0.000000          0.000000          3.000000      0.000000   
25%            0.000000          0.000000          6.000000      0.000000   
50%            0.000000          0.000000          8.000000      0.000000   
75%            0.000000          1.000000          9.000000      0.000000   
max           76.000000         21.000000         16.000000      1.000000   

       dosage_change  
count   96447.000000  
mean        0.288262  
std         0.488613  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         1.000000  
max         4.000000

In [194]:
# code the non-numeric values
new_df['change'] = new_df['change'].replace('Ch', 1)
new_df['change'] = new_df['change'].replace('No', 0)

new_df['gender'] = new_df['gender'].replace('Male', 1)
new_df['gender'] = new_df['gender'].replace('Female', 0)

new_df['diabetesMed'] = new_df['diabetesMed'].replace('Yes', 1)
new_df['diabetesMed'] = new_df['diabetesMed'].replace('No', 0)

In [195]:
new_df.head()

encounter_id  patient_nbr             race gender  age  admission_type_id  \
1        149190     55629189        Caucasian      0   15                  1   
2         64410     86047875  AfricanAmerican      0   25                  1   
3        500364     82442376        Caucasian      1   35                  1   
4         16680     42519267        Caucasian      1   45                  1   
5         35754     82637451        Caucasian      1   55                  2   

   discharge_disposition_id  admission_source_id  time_in_hospital  \
1                         1                    7                 3   
2                         1                    7                 2   
3                         1                    7                 2   
4                         1                    7                 1   
5                         1                    2                 3   

   num_lab_procedures  ...  insulin  glyburide-metformin  glipizide-metformin  \
1                  59  ...       Up                   No                   No   
2                  11  ...       No                   No                   No   
3                  44  ...       Up                   No                   No   
4                  51  ...   Steady                   No                   No   
5                  31  ...   Steady                   No                   No   

   glimepiride-pioglitazone  metformin-rosiglitazone metformin-pioglitazone  \
1                        No                       No                     No   
2                        No                       No                     No   
3                        No                       No                     No   
4                        No                       No                     No   
5                        No                       No                     No   

  change diabetesMed  readmitted dosage_change  
1      1           1           0             1  
2      0           1           0             0  
3      1           1           0             1  
4      1           1           0             0  
5      0           1           0             0  

[5 rows x 48 columns]

In [200]:
new_df["A1Cresult"].value_counts()

None    80394
>8       7567
Norm     4806
>7       3680
Name: A1Cresult, dtype: int64

In [225]:
new_df['A1Cresult'] = new_df['A1Cresult'].replace('>8', 1)
new_df['A1Cresult'] = new_df['A1Cresult'].replace('>7', 1)
new_df['A1Cresult'] = new_df['A1Cresult'].replace('Norm', 0)
new_df['A1Cresult'] = new_df['A1Cresult'].replace('None', 2)

In [226]:
new_df["A1Cresult"].value_counts()

2    80394
1    11247
0     4806
Name: A1Cresult, dtype: int64

In [207]:
new_df["number_inpatient"].value_counts()

0     63708
1     18720
2      7237
3      3258
4      1564
5       790
6       471
7       262
8       142
9       111
10       58
11       48
12       33
13       18
14       10
15        8
16        5
19        2
18        1
21        1
Name: number_inpatient, dtype: int64

In [208]:
new_df.columns

Index(['encounter_id', 'patient_nbr', 'race', 'gender', 'age',
       'admission_type_id', 'discharge_disposition_id', 'admission_source_id',
       'time_in_hospital', 'num_lab_procedures', 'num_procedures',
       'num_medications', 'number_outpatient', 'number_emergency',
       'number_inpatient', 'diag_1', 'diag_2', 'diag_3', 'number_diagnoses',
       'max_glu_serum', 'A1Cresult', 'metformin', 'repaglinide', 'nateglinide',
       'chlorpropamide', 'glimepiride', 'acetohexamide', 'glipizide',
       'glyburide', 'tolbutamide', 'pioglitazone', 'rosiglitazone', 'acarbose',
       'miglitol', 'troglitazone', 'tolazamide', 'examide', 'citoglipton',
       'insulin', 'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted',
       'dosage_change'],
      dtype='object')

In [224]:
new_df["gender"].value_counts()

0                  52006
1                  44440
Unknown/Invalid        1
Name: gender, dtype: int64

In [227]:
new_df['gender'] = new_df['gender'].replace('Unknown/Invalid', 0)

In [228]:
new_df["gender"].value_counts()

0    52007
1    44440
Name: gender, dtype: int64

I know that I have to map the admission_type_id, discharge_disposition_id and admission_source_id I will do it manually right now just to run my MVP my replacing my Null/not mapped to the same category ( and other categories if fast)

In [230]:
# Let's start with my admission_type_id
new_df['admission_type_id'] = new_df['admission_type_id'].replace(6, 5)
new_df['admission_type_id'] = new_df['admission_type_id'].replace(8, 5)
new_df['admission_type_id'] = new_df['admission_type_id'].replace(7, 1)
new_df['admission_type_id'] = new_df['admission_type_id'].replace(2, 1)

# Discharge_disposition_id
new_df['discharge_disposition_id'] = new_df['discharge_disposition_id'].replace(26, 18)
new_df['discharge_disposition_id'] = new_df['discharge_disposition_id'].replace(25, 18)

#Admission_source_id

new_df['admission_source_id'] = new_df['admission_source_id'].replace(17, 9)
new_df['admission_source_id'] = new_df['admission_source_id'].replace(15, 9)
new_df['admission_source_id'] = new_df['admission_source_id'].replace(20, 9)
new_df['admission_source_id'] = new_df['admission_source_id'].replace(21, 9)

In [231]:
new_df["max_glu_serum"].value_counts()

None    91358
Norm     2510
>200     1404
>300     1175
Name: max_glu_serum, dtype: int64

In [232]:
new_df['max_glu_serum'] = new_df['max_glu_serum'].replace(">200", 1)
new_df['max_glu_serum'] = new_df['max_glu_serum'].replace(">300", 1)
new_df['max_glu_serum'] = new_df['max_glu_serum'].replace("Norm", 0)
new_df['max_glu_serum'] = new_df['max_glu_serum'].replace("None", 2)

In [233]:
new_df["max_glu_serum"].value_counts()

2    91358
1     2579
0     2510
Name: max_glu_serum, dtype: int64

In [257]:
new_df["A1Cresult"]. value_counts()

2    80394
1    11247
0     4806
Name: A1Cresult, dtype: int64

In [258]:
# Creating dummmies for the race column

df_mvp =pd.get_dummies(new_df, columns=['race'], drop_first=True)
df_mvp.head()


encounter_id  patient_nbr  gender  age  admission_type_id  \
1        149190     55629189       0   15                  1   
2         64410     86047875       0   25                  1   
3        500364     82442376       1   35                  1   
4         16680     42519267       1   45                  1   
5         35754     82637451       1   55                  1   

   discharge_disposition_id  admission_source_id  time_in_hospital  \
1                         1                    7                 3   
2                         1                    7                 2   
3                         1                    7                 2   
4                         1                    7                 1   
5                         1                    2                 3   

   num_lab_procedures  num_procedures  ...  metformin-rosiglitazone  \
1                  59               0  ...                       No   
2                  11               5  ...                       No   
3                  44               1  ...                       No   
4                  51               0  ...                       No   
5                  31               6  ...                       No   

   metformin-pioglitazone  change  diabetesMed readmitted dosage_change  \
1                      No       1            1          0             1   
2                      No       0            1          0             0   
3                      No       1            1          0             1   
4                      No       1            1          0             0   
5                      No       0            1          0             0   

  race_Asian  race_Caucasian  race_Hispanic  race_Other  
1          0               1              0           0  
2          0               0              0           0  
3          0               1              0           0  
4          0               1              0           0  
5          0               1              0           0  

[5 rows x 51 columns]

# MVP: Logistic Regression

In [259]:
x_mvp = df_mvp.drop(['encounter_id', 'patient_nbr','readmitted','diag_1', 'diag_2', 'diag_3','metformin', 'repaglinide', 'nateglinide',
       'chlorpropamide', 'glimepiride', 'acetohexamide', 'glipizide',
       'glyburide', 'tolbutamide', 'pioglitazone', 'rosiglitazone', 'acarbose',
       'miglitol', 'troglitazone', 'tolazamide', 'examide', 'citoglipton',
       'insulin', 'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone'], axis = 1)

y_mvp = df_mvp["readmitted"]

In [260]:
x_mvp.columns

Index(['gender', 'age', 'admission_type_id', 'discharge_disposition_id',
       'admission_source_id', 'time_in_hospital', 'num_lab_procedures',
       'num_procedures', 'num_medications', 'number_outpatient',
       'number_emergency', 'number_inpatient', 'number_diagnoses',
       'max_glu_serum', 'A1Cresult', 'change', 'diabetesMed', 'dosage_change',
       'race_Asian', 'race_Caucasian', 'race_Hispanic', 'race_Other'],
      dtype='object')

In [261]:
from sklearn import  model_selection
from sklearn import linear_model, metrics
x_train_mvp, x_testval_mvp, y_train_mvp, y_testval_mvp = model_selection.train_test_split(
    x_mvp,
    y_mvp,
    test_size=.5,
    stratify=y_mvp,      # stratify ensures that each split has the same number in each class
)

x_test_mvp, x_val_mvp, y_test_mvp, y_val_mvp = model_selection.train_test_split(
    x_testval_mvp,
    y_testval_mvp,
    test_size=.5,
    stratify=y_testval_mvp,      # stratify ensures that each split has the same number in each class
)

In [262]:
print(x_train_mvp.shape,y_train_mvp.shape)
print(x_test_mvp.shape, y_test_mvp.shape)

(48223, 22) (48223,)
(24112, 22) (24112,)


In [265]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import precision_score, recall_score, precision_recall_curve,f1_score, fbeta_score
from sklearn.model_selection import GridSearchCV
import numpy as np
logit = LogisticRegression(C= 0.95)
logit.fit(x_train_mvp, y_train_mvp)
y_pred_mvp = logit.predict(x_val_mvp)

print("Training: {:6.2f}%".format(100*logit.score(x_train_mvp, y_train_mvp)))
print("Test set: {:6.2f}%".format(100*logit.score(x_test_mvp, y_test_mvp)))

/anaconda3/envs/metis/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Training:  88.47%
Test set:  88.55%


In [266]:
from sklearn.metrics import classification_report, confusion_matrix
print(confusion_matrix(y_test_mvp,y_pred_mvp))
print(classification_report(y_test_mvp,y_pred_mvp))

[[21281    65]
 [ 2761     5]]
              precision    recall  f1-score   support

           0       0.89      1.00      0.94     21346
           1       0.07      0.00      0.00      2766

    accuracy                           0.88     24112
   macro avg       0.48      0.50      0.47     24112
weighted avg       0.79      0.88      0.83     24112

